duo.c

```diff
@@ -148,19 +175,22 @@ _sign_request(struct duo_ctx *ctx, const char *method, const char *uri,
         strftime(date, sizeof(date), "%a, %d %b %Y %T %z", localtime(&t));
         
         /* Generate signature over the canonicalized request */
-        HMAC_CTX_init(&hmac);
-	HMAC_Init(&hmac, ctx->skey, strlen(ctx->skey), EVP_sha1());
-        HMAC_Update(&hmac, (u_char *)date, strlen(date));
-        HMAC_Update(&hmac, (u_char *)"\n", 1);
-        HMAC_Update(&hmac, (u_char *)method, strlen(method));
-        HMAC_Update(&hmac, (u_char *)"\n", 1);
-        HMAC_Update(&hmac, (u_char *)ctx->host, strlen(ctx->host));
-        HMAC_Update(&hmac, (u_char *)"\n", 1);
-        HMAC_Update(&hmac, (u_char *)uri, strlen(uri));
-        HMAC_Update(&hmac, (u_char *)"\n", 1);
-        HMAC_Update(&hmac, (u_char *)qs, strlen(qs));
-	HMAC_Final(&hmac, MD, NULL);
-	HMAC_CTX_cleanup(&hmac);
+        if ((hmac = HMAC_CTX_new()) == NULL) {
+           return (NULL);
+        }
+
+	HMAC_Init_ex(hmac, ctx->skey, strlen(ctx->skey), EVP_sha1(), NULL);
```

Comment:
My only remaining concern was that HMAC_Init_ex doesn't appear to be on anything less than 1.0.2 but anything less than 1.0.2 is no longer supported by OpenSSL so now I'm good with this change.